qa/L0_model_update/instance_update_test.py

```diff
@@ -0,0 +1,310 @@
+                                        "instance_init_del")
+            update_thread.result()
+            update_end_time = time.time()
+            infer_thread.result()
+            infer_end_time = time.time()
+        infer_time = infer_end_time - infer_start_time
+        update_time = update_end_time - update_start_time
+        # Adding a new instance does not depend on existing instances, so the
+        # ongoing inference should not block the update.
+        self.assertGreaterEqual(infer_time, 10.0, "Invalid infer time")
+        self.assertLess(update_time, 5.0, "Update blocked by infer")
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 2)
+        with self.assertRaises(InferenceServerException):
+            self.__triton.infer("instance_init_del", self.__get_inputs((4,)))
+
+    def test_infer_while_updating(self):
+        # Load model
+        update_instance_group("{\ncount: 1\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 1)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
+        # Infer while adding 1 instance
+        set_initialize_delay(10)
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            update_start_time = time.time()
+            update_thread = pool.submit(self.__triton.load_model,
+                                        "instance_init_del")
+            time.sleep(2)  # make sure update has started
+            infer_start_time = time.time()
+            infer_thread = pool.submit(self.__triton.infer, "instance_init_del",
+                                       self.__get_inputs((1,)))
+            infer_thread.result()
+            infer_end_time = time.time()
+            update_thread.result()
+            update_end_time = time.time()
+        update_time = update_end_time - update_start_time
+        infer_time = infer_end_time - infer_start_time
+        # Waiting on new instance creation should not block inference on
+        # existing instances.
+        self.assertGreaterEqual(update_time, 10.0, "Invalid update time")
+        self.assertLess(infer_time, 5.0, "Infer blocked by update")
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
```

Comment:
Same as above. These things can go to a utility function.

Reply:
Updated: [Refactor instance update test](https://github.com/triton-inference-server/server/pull/5635/commits/f012b0474ac4b1465e432e72f3c4f24cb5cb6b3c)